### PR TITLE
refactor(plugin-server): move buffer as first step of event pipeline & more

### PR DESCRIPTION
**Pull request by macobo**
## Problem

This PR started out as a solution for https://github.com/PostHog/posthog/issues/10192.

However, I removed the part on updating before plugins as we haven't shipped versioning persons by version yet. In a later PR can undo this commit: [`Undo radical change to event pipeline - will re-add it later!`](https://github.com/PostHog/posthog/pull/10360/commits/f541b58242d827e68db5e507425caf5e225e5fed).

## Changes

This PR:
- Moves buffer check as the first step of the event pipeline
- Extracts person updating as a separate step in the pipeline
- Reduces the number of redundant `person` fetches we do by a lot by caching/forwarding results
- Contains small misc changes for https://github.com/PostHog/posthog/issues/10192

Operationally very little should change besides:
- Changing the format of the buffer topic
- Reducing load on postgresql for person fetches

## How did you test this code?

See tests - this has extensive coverage.

### CODE REVIEW

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (tiina303) [inline plugin-server/src/worker/ingestion/person-state.ts:95]: naming is slightly confusing as this can return undefined especially at the end return

File: plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts
Comment (tiina303) [inline plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts:28]: Since we always need a person in the future steps and how it's better if person lookups happen in order (person properties will more naturally change on events), consider looking up the person from redis cache here & make prepareEventStep not need to deal with it being `undefined` (i.e. move https://github.com/PostHog/posthog/blob/f541b58242d827e68db5e507425caf5e225e5fed/plugin-server/src/worker/ingestion/process-event.ts#L221-L228 here).

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (tiina303) [inline plugin-server/src/worker/ingestion/person-state.ts:180]: I'm confused it says that we always fetch the person and we should (based on current property updates), but the code uses `this.person`, which potentially came from merging, which could be in a different kafka partition, i.e. processed in parallel to updates to this person.

**Review by tiina303 — COMMENTED** [review]
flyby comments

File: plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts:28]: The redis cache is IMO sort of unneeded and should be removed completely - it didn't do much before given we were calling `fetchPerson` up to 6 times per event.

Now for most vanilla events we only fetch it once in the `emitToBuffer` step.

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (macobo) [reply plugin-server/src/worker/ingestion/person-state.ts:180]: Comment is out of date, will remove.

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/person-state.ts:181]: also what's the impact of using this cached person here? we will be more subject to races, no? or what's the analysis that led to the decision of using the cache here? in order ingestion? (even though races are possible with identify)

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/person-state.ts:289]: I'm aware of the context but this comment doesn't help much if the reader doesn't have this context. I mean as in why this is `TRICKY`

File: plugin-server/src/worker/ingestion/event-pipeline/4-prepareEventStep.ts
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/event-pipeline/4-prepareEventStep.ts:28]: `preIngestionEvent as IngestionEvent` is a bit confusing - a comment for now might be sufficient

**Review by yakkomajuri — COMMENTED** [review]
good stuff - not as bad to review as it seemed

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (macobo) [reply plugin-server/src/worker/ingestion/person-state.ts]: This has been removed.

File: plugin-server/src/worker/ingestion/event-pipeline/4-prepareEventStep.ts
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/4-prepareEventStep.ts:28]: Not sure what to comment on this - ingestion already happened, so we can treat this as an IngestionEvent?

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (macobo) [reply plugin-server/src/worker/ingestion/person-state.ts:289]: Because we're basically modifying 'global' state (for this class) instead of relying on return values. It's bad code.

However the whole `handleIdentify` logic is so convoluted that I didn't want to get into extensive refactoring and risk breaking it. Instead I'm relying on extensive code coverage.

File: plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts
Comment (tiina303) [reply plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts:28]: sure we can remove the cache, my comment was more about the fact that lets make the person step return the person always and later steps not need to think about it.

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (tiina303) [reply plugin-server/src/worker/ingestion/person-state.ts]: I think if we use person we get from merging there could be other updates which we would be overriding if we drop this, because of the way we do property updates, that we should solve that problem as a whole, so it's fine to make this a bit worse ... probably ... just be aware that this can cause some props updates to be lost.

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (tiina303) [reply plugin-server/src/worker/ingestion/person-state.ts:181]: > I have a counter-question: Why would we want to refetch persons multiple times in this pipeline? I can see it being no less prone to races and leads directly worse performance. 

Agreed it's not good for perf nor Postgres load.

> The only reason we were consistently fetching these persons didn't have anything to do with correctness but just a side-effect of poorly tested code

I wish this was true, unfortunately it is tied to correctness here due to the way we do props updates, but we might say yolo and ignore the potential races and get the perf win (fine by me, we should re-design props updating, but I don't have the proposal written up yet).

Here's an example of what can happen now:

Let's say we have two executions running in parallel: 
1. A - which does an alias
2. B which does a random property update. 
 
Note that due to the fact that A does an alias it's possible that these are in different partitions, which means could be processed in parallel even after our recent changes to break batches up by distinctId.

Imagine the following execution flow in time: 
1. A reached here `const personFound = this.person` and uses the person they merged from the beginning. 
2. B reaches here, fetches the person (if not with `for update` we're prone to this problem here anyway)
5. B updates some properties & completes execution
6. A computes the props update agains the person from step (1)/earlier & overrides changes B made (as we just dump the whole json that happens even if different props were changed).

File: plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/3-processPersonsStep.ts:28]: Not sure I understand. The `1-emitBufferStep` already always fetches person and it's used later in the pipeline.

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (macobo) [reply plugin-server/src/worker/ingestion/person-state.ts:181]: @tiina303 the exact same race condition still exists in the previous version as the alias could happen between fetchPerson in `updateProperties` and the actual updating.

By re-fetching I posit we were not actually solving any races - just making them harder to detect.

Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/person-state.ts:181]: Would having a metric like [this](https://github.com/PostHog/posthog/pull/10423) help us feel more comfortable releasing this @tiina303?

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/person-state.ts:289]: Was a comment on the comment rather than what it's doing, but either way, no big deal

**Review by yakkomajuri — COMMENTED** [review]
I'm happy with this but would recommend ironing things out with @tiina303 first and maybe just get my PR in with the metric for a bit before we merge this so we at least have a frame for discussion.

Effectively happy to stamp once we have a metric that helps us get a sense of how much worse (or not) we might be making this problem.

I'm also aware that even the current approach is subject to races, even though it tries a bit to minimize them

**Comment by macobo** [comment]
>I'm also aware that even the current approach is subject to races, even though it tries a bit to minimize them

I'm gonna call "bullshit" on this - we were inefficient due to bad software design not due to higher level of planning/conscious decisions.

File: plugin-server/src/worker/ingestion/person-state.ts
Comment (tiina303) [reply plugin-server/src/worker/ingestion/person-state.ts:181]: I tried to say twice in my comment that it's good to be merged & didn't request changes. So let me repeat: it's good to go in once Yakko or Michael approved. Sorry if that wasn't clear, I'll be more careful about my wording "yolo" in the future. I did want to call out there this makes things a bit worse & provided an example. The reason I commented in the first place was so it would be faster for us to find / troubleshoot if after merging we see problems.

**Review by yakkomajuri — APPROVED** [review]
let's get this in tomorrow morning. I added in a metric that's not perfect but should give us some overview of things

**Pull request merged**
